README.md
[![Unix build](https://img.shields.io/github/actions/workflow/status/Tieske/corowatch/unix_build.yml?branch=master&label=Unix%20build&logo=linux)](https://github.com/Tieske/corowatch/actions/workflows/unix_build.yml)
[![Coveralls code coverage](https://img.shields.io/coveralls/github/Tieske/corowatch?logo=coveralls)](https://coveralls.io/github/Tieske/corowatch)
[![Lint](https://github.com/Tieske/corowatch/workflows/Lint/badge.svg)](https://github.com/Tieske/corowatch/actions/workflows/lint.yml)
[![SemVer](https://img.shields.io/github/v/tag/Tieske/corowatch?color=brightgreen&label=SemVer&logo=semver&sort=semver)](CHANGELOG.md)

# corowatch

Lua module to watch coroutine usage and kill a coroutine if it fails to yield in a timely manner.

## License and copyright

See [LICENSE.md](LICENSE.md)

## Documentation

See [online documentation](https://Tieske.github.io/corowatch/)

## Changelog

See [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# CHANGELOG

#### Releasing new versions

- create a release branch
- update the changelog below
- update version and copyright-years in `./LICENSE` and `./src/corowatch.lua` (in doc-comments
  header, and in module constants)
- create a new rockspec and update the version inside the new rockspec:<br/>
  `cp corowatch-scm-1.rockspec ./rockspecs/corowatch-X.Y.Z-1.rockspec`
- test: run `make test` and `make lint`
- clean and render the docs: run `make clean` and `make docs`
- commit the changes as `release X.Y.Z`
- push the commit, and create a release PR
- after merging tag the release commit with `version_X.Y.Z`
- upload to LuaRocks:<br/>
  `luarocks upload ./rockspecs/corowatch-X.Y.Z-1.rockspec --api-key=ABCDEFGH`
- test the newly created rock:<br/>
  `luarocks install corowatch`


### Version 1.1, released 30-Dec-2022

- feat: make `hookcount` configurable
- fix: in Lua 5.1, sethook should be called without "coro" for the main thread
- fix: make pack/unpack nil-safe

### Version 1.0, released 04-Feb-2014

- no automatic monkey patching of globals anymore

### Version 0.2, released 07-Apr-2013

- fixed debughook settings, improved performance

### Version 0.1, released 04-Apr-2013

- initial release


spec/corowatch_spec.lua
local corowatch

if jit then -- JITted code will not call debug hooks, so disable it
  jit.off()
end

describe("testing the corowatch module", function()

  setup(function()
    _G._TEST = true
    corowatch = require("corowatch")
    corowatch.export(_G)
  end)

  teardown(function()
  end)

  before_each(function()
  end)

  after_each(function()
    collectgarbage()  -- make sure to collect weak references in register table
  end)

  it("Checks exported methods", function()
    local coroutine = {}
    local debug = {}
    local globals = {coroutine = coroutine, debug = debug}
    corowatch.export(globals)
    assert.are_equal(corowatch.wrapf,coroutine.wrapf)
    assert.are_equal(corowatch.create,coroutine.create)
    assert.are_equal(corowatch.yield,coroutine.yield)
    assert.are_equal(corowatch.resume,coroutine.resume)
    assert.are_equal(corowatch.wrap,coroutine.wrap)
    --assert.are_equal(corowatch.running,coroutine.running)
    assert.are_equal(corowatch.status,coroutine.status)
    assert.are_equal(corowatch.sethook,debug.sethook)
  end)

  it("Checks a coroutine being registered when watched", function()
    local kill_expect, warn_expect, cb_expect, hookcount_expect = 2, 1, function() end, 123
    assert.is_nil(next(corowatch._register))  -- check register is empty when starting
    local c = corowatch.watch(coroutine.create(function() end), kill_expect, warn_expect, cb_expect, hookcount_expect)
    local w = corowatch._register[c]
    assert.not_is_nil(w)
    assert.are_equal(kill_expect, w.tkilllimit)
    assert.are_equal(warn_expect, w.twarnlimit)
    assert.are_equal(cb_expect, w.cb)
    assert.are_equal(hookcount_expect, w.hookcount)
    assert.is_nil(w.warned)
    assert.is_nil(w.errmsg)
    assert.is_nil(w.killtime)
    assert.is_nil(w.warntime)
    assert.are_equal(debug.gethook(c), w.hook)
  end)

  pending("Checks a watch entry when a coroutine is resumed", function()
    local kill_expect, warn_expect, cb_expect = 2, 1, function() end
    assert.is_nil(next(corowatch._register))  -- check register is empty when starting
    local w = {}
    local f = function()
        -- copy contents of watch table now the coro is running
        for k,v in pairs(corowatch._register[coroutine.running()]) do
          w[k] = v
        end
      end
    local c = corowatch.watch(coroutine.create(f), kill_expect, warn_expect, cb_expect)
    local t = corowatch.gettime()
    coroutine.resume(c)
    assert.are_equal(kill_expect, w.tkilllimit)
    assert.are_equal(warn_expect, w.twarnlimit)
    assert.are_equal(cb_expect, w.cb)
    assert.is_nil(w.warned)
    assert.is_nil(w.errmsg)
    assert.are_equal(t, w.killtime - kill_expect)
    assert.are_equal(t, w.warntime - warn_expect)
    assert.are_equal(debug.gethook(c), w.hook)
  end)

  it("Checks a watch entry when a coroutine is yielded", function()
    local kill_expect, warn_expect, cb_expect = 2, 1, function() end
    assert.is_nil(next(corowatch._register))  -- check register is empty when starting
    local f = function()
        coroutine.yield(123, "abc")
      end
    local c = corowatch.watch(coroutine.create(f), kill_expect, warn_expect, cb_expect)
    local ok, one, two = coroutine.resume(c)
    local w = corowatch._register[c]
    assert.is_true(ok)
    assert.are_equal(123, one)
    assert.are_equal("abc", two)
    assert.are_equal(kill_expect, w.tkilllimit)
    assert.are_equal(warn_expect, w.twarnlimit)
    assert.are_equal(cb_expect, w.cb)
    assert.is_nil(w.warned)
    assert.is_nil(w.errmsg)
    assert.is_nil(w.killtime)
    assert.is_nil(w.warntime)
    assert.are_equal(debug.gethook(c), w.hook)
  end)

  it("Checks errors when creating a watch", function()
    local c = corowatch.watch(coroutine.create(function() end), 2, 1, function() end)
    assert.has_error(function()
          corowatch.watch(c)  -- cannot add another watch
        end)
    assert.has_error(function() -- when warnlimit, also callback must be provided
          c = corowatch.watch(coroutine.create(function() end), 2, 1)
        end)
    assert.has_error(function() -- either warnlimit or killlimit must be provided
          c = corowatch.watch(coroutine.create(function() end))
        end)
    assert.has_error(function() -- warnlimit must be smaller than killimit
          c = corowatch.watch(coroutine.create(function() end), 1, 2)
        end)
    assert.has_error(function() -- hookcount must be >= 1
          c = corowatch.watch(coroutine.create(function() end), 1, 2, -1)
        end)
  end)

  it("checks that a coroutine created from a watched coroutine is also watched", function()
    local kill_expect, warn_expect, cb_expect, hookcount_expect = 2, 1, function() end, 123
    local first, second
    local f = function()
      second = coroutine.create(function() end)
    end
    first = corowatch.watch(coroutine.create(f), kill_expect, warn_expect, cb_expect, hookcount_expect)
    coroutine.resume(first)
    -- now check whether 'second' inherited the watch values
    local w = corowatch._register[second]
    assert.not_is_nil(w)
    assert.are_equal(kill_expect, w.tkilllimit)
    assert.are_equal(warn_expect, w.twarnlimit)
    assert.are_equal(cb_expect, w.cb)
    assert.are_equal(hookcount_expect, w.hookcount)
    assert.is_nil(w.warned)
    assert.is_nil(w.errmsg)
    assert.is_nil(w.killtime)
    assert.is_nil(w.warntime)
    assert.are_equal(debug.gethook(second), w.hook)
  end)

  it("checks that a timedout coroutine gets status 'dead'", function()
    local kill_expect, warn_expect, cb_expect = 2, 1, function() end
    local f = function() end
    local c = corowatch.watch(coroutine.create(f), kill_expect, warn_expect, cb_expect)
    assert.are_equal("suspended", coroutine.status(c))
    local w = corowatch._register[c]
    w.errmsg = "just some error that is not nil"
    assert.are_equal("dead", coroutine.status(c))
  end)

  it("checks killing and warning time", function()
    local kill_expect, warn_expect = .5, .25
    local tt = {}
    local warncount = 1
    local killcount = 0
    local cb = function(cbt)
        if cbt == "kill" then
          killcount = killcount + 1
          return
        end
        warncount = warncount + 1
        tt[warncount] = corowatch.gettime()
        if warncount < 3 then return true end  -- continue
      end
    local f = function()
        -- just do something silly in a loop
        while true do
          local i = 0
          i = i + 1
          i = i - 1   -- luacheck: ignore
        end
      end
    local c = corowatch.watch(coroutine.create(f), kill_expect, warn_expect, cb)
    tt[1] = corowatch.gettime()
    local ok, msg = coroutine.resume(c)
    tt[4] = corowatch.gettime()
    local w = corowatch._getwatch(c)
    assert.is_false(ok)
    assert.are_equal(msg, w.errmsg)
    assert.are_equal(3, warncount)
    assert.are_equal(1, killcount)
    assert(tt[2]-tt[1]-warn_expect < 0.1, "expected time difference to be less than 0.1 seconds")
    assert(tt[3]-tt[2]-warn_expect < 0.1, "expected time difference to be less than 0.1 seconds")
    assert(tt[4]-tt[3]-kill_expect < 0.1, "expected time difference to be less than 0.1 seconds")
    assert.are_equal("dead", coroutine.status(c))
  end)

  it("checks that coroutine.wrap() works as expected with a watched coroutine", function()
    local f = function()
        local t = corowatch.gettime() + 3
        while t > corowatch.gettime() do
          -- do something silly in a loop
          local i = 0
          i = i + 1
          i = i - 1   -- luacheck: ignore
        end
      end
    local wf
    -- create/run a watched coroutine from which 'wrap' is called
    coroutine.resume(corowatch.watch(coroutine.create(function()
          wf = coroutine.wrap(f)
        end), 0.25))
    -- now the coroutine embedded in wf should be watched
    local one, two = wf()
    assert.is_false(one)  -- failure report
    assert.is_string(two) -- error message
  end)


end)


examples/performance.lua

--package.path = "C:\\Users\\Thijs\\Dropbox\\Lua projects\\corowatch\\src\\?.lua;"..package.path
local corowatch = require("corowatch")
corowatch.export(_G)

local res = 1

local testfunc = function()
  for n = 1,2000000 do
    res=math.sin(n/1000)
  end
  return res
end

local function test1()
  collectgarbage()
  collectgarbage()
  collectgarbage("stop")
  local t1 = corowatch.gettime()
  testfunc()
  t1 = corowatch.gettime() - t1
  collectgarbage("restart")
  collectgarbage()
  collectgarbage()
  return t1
end

local function test2(wrapper)
  -- wrapper = coroutine.wrap, or coroutine.wrapf
  collectgarbage()
  collectgarbage()
  collectgarbage("stop")
  local t2 = corowatch.gettime()
  wrapper(testfunc, 10001, 10000, function() end)()
  t2 = corowatch.gettime() - t2
  collectgarbage("restart")
  collectgarbage()
  collectgarbage()
  return t2
end

-- warm up
for n = 1,3 do
  test1()
  test2(coroutine.wrap)
  test2(coroutine.wrapf)   -- luacheck: ignore
end

-- run test
local t0, t1, t2, iter = 0,0,0,10
for n = 1,iter do
  t0=test1() + t0
  t1=test2(coroutine.wrap) + t1
  t2=test2(coroutine.wrapf) + t2   -- luacheck: ignore
end
t0=t0/iter  -- main loop
t1=t1/iter  -- coroutine
t2=t2/iter  -- protected coroutine

print("Mainloop :",t0)
print("Coroutine:",t1)
print("Corowatch:",t2)
print("corowatch is " .. math.floor((t2-t1)/t1 *100) .. "% slower than unprotected")
print("coroutine is " .. math.floor((t0-t1)/t0 *100) .. "% faster than the main loop")



src/corowatch.lua
---------------------------------------------------------------------------------------
-- Module to watch coroutine executiontime. Coroutines running too long without
-- yielding can be killed to prevent them from locking the Lua state.
-- The module uses `LuaSocket` to get the time (`socket.gettime` function). If you
-- do not want that, override the `coroutine.gettime` method with your own
-- implementation.
--
-- @copyright Copyright (c) 2013-2022 Thijs Schreijer
-- @author Thijs Schreijer
-- @license MIT, see `LICENSE.md`.
-- @name corowatch
-- @class module

local M = {}    -- module table
M._VERSION = "1.1"
M._COPYRIGHT = "Copyright (c) 2013-2022 Thijs Schreijer"
M._DESCRIPTION = "Lua module to watch coroutine usage and kill a coroutine if it fails to yield in a timely manner."


local corocreate = coroutine.create
local cororesume = coroutine.resume
local cororunning = coroutine.running
local corostatus = coroutine.status
local corowrap = coroutine.wrap
local coroyield = coroutine.yield
local _sethook = debug.sethook
local traceback = debug.traceback
local watch, resume, create
local default_hookcount = 10000


local pack, unpack do -- pack/unpack to create/honour the .n field for nil-safety
  local _unpack = _G.table.unpack or _G.unpack
  function pack (...) return { n = select('#', ...), ...} end
  function unpack(t, i, j) return _unpack(t, i or 1, j or t.n or #t) end
end

-- create watch register; table indexed by coro with watch properties
local register = setmetatable({},{__mode = "k"})  -- set weak keys
-- register element = {
--   twarnlimit = when to warn, seconds
--   tkilllimit = when to kill, seconds
--   cb = callback, function
--   warned = boolean; was cwwarn called already?
--   killtime = point in time after which to kill the coroutine
--   warntime = point in time after which to warn
--   errmsg = errormessage if timedout, so if set, the coro should be dead
--   hook = function; the debughook in use
--   debug = debug.traceback() at the time of protecting
--   hookcount = the count setting for the debughook
-- }


local mainthread = {}  -- for 5.1 where there is no main thread access, running() returns nil
local function sethook(coro, ...)  -- compatibility for Lua 5.1
  if coro == mainthread then
    return _sethook(...)
  end
  return _sethook(coro, ...)
end

-- Gets the entry for the coro from the coroutine register.
-- @return the coroutine entry from the register
local getwatch = function(coro)
  return register[coro or cororunning() or mainthread]
end


-- Debughook function, to check for timeouts
local checkhook = function()
  local e = getwatch()
  if not e then return end  -- not being watched
  local t = M.gettime()
  if e.errmsg then
    -- the coro is tagged with an error. This means that somewhere the error
    -- was thrown, but the coro didn't die (probably a pcall in between). The coro should have died
    -- but didn't so it is in an 'undetermined' state. So kill again.
    error(e.errmsg,2)
  elseif not e.warned and e.warntime and e.warntime < t then
    -- warn now
    if e.cb("warn") then
      -- returned positive, so reset timeouts
      if e.tkilllimit then e.killtime = t + e.tkilllimit end
      if e.twarnlimit then e.warntime = t + e.twarnlimit end
    else
      e.warned = true
    end
  elseif e.killtime and e.killtime < t then
    if e.cb then e.cb("kill") end
    -- run hook now every instruction, to kill again and again if it persists (pcall/cboundary)
    sethook(cororunning(), e.hook, "", 1)
    -- kill now
    e.errmsg = "Coroutine exceeded its allowed running time of "..tostring(e.tkilllimit).." seconds, without yielding"
    if e.debug then
      e.errmsg = e.errmsg ..
       "\n============== traceback at coroutine creation time ====================\n" ..
       e.debug ..
       "\n======================== end of traceback =============================="
    end
    error(e.errmsg, 2)
  end
end

-- Creates an entry in the coroutine register. If it already exists, existing values
-- will be overwritten (in the existing entry) with the new values.
-- @param coro coroutine for which to create the entry
-- @return the entry created
local createwatch = function(coro, tkilllimit, twarnlimit, cb, hookcount)
  coro = coro or cororunning() or mainthread
  hookcount = hookcount or default_hookcount
  local entry = register[coro] or {}
  entry.tkilllimit = tkilllimit
  entry.twarnlimit = twarnlimit
  entry.cb = cb
  entry.hook = checkhook
  entry.debug = traceback()
  entry.hookcount = hookcount
  sethook(coro, entry.hook, "", hookcount)
  register[coro] = entry
  return entry
end

---------------------------------------------------------------------------------------
-- returns current time in seconds. If not overridden, it will require `luasocket` and use
-- `socket.gettime` to get the current time.
M.gettime = function()
  M.gettime = require("socket").gettime
  return M.gettime()
end

---------------------------------------------------------------------------------------
-- Protects a coroutine from running too long without yielding.
-- The callback has 1 parameter (string value being either "warn" or "kill"), but runs
-- on the coroutine that is subject of the warning. If the "warn" callback returns a
-- truthy value (neither `false`, nor `nil`) then the timeouts for kill and warn limits
-- will be reset (buying more time for the coroutine to finish its business).
--
-- The `hookcount` default of 10000 will ensure offending coroutines are caught with
-- limited performance impact. To better narrow down any offending code that takes too long,
-- this can be set to a lower value (eg. set it to 1, and it will break right after
-- the instruction that tripped the limit). But the smaller the value, the higher the
-- performance cost.
--
-- NOTE: the callback runs inside a debughook.
-- @tparam coroutine|nil coro coroutine to be protected, defaults to the currently running routine
-- @tparam number|nil tkilllimit time in seconds it is allowed to run without yielding
-- @tparam number|nil twarnlimit time in seconds it is allowed before `cb` is called
-- (must be smaller than `tkilllimit`)
-- @tparam function|nil cb callback executed when the kill or warn limit is reached.
-- @tparam[opt=10000] number hookcount the hookcount to use (every `x` number of VM instructions check the limits)
-- @return coro
M.watch = function(coro, tkilllimit, twarnlimit, cb, hookcount)
  if getwatch(coro) then
    error("Cannot create a watch, there already is one")
  end
  assert(tkilllimit or twarnlimit, "Either kill limit or warn limit must be provided")
  if twarnlimit ~= nil then
    assert(type(twarnlimit) == "number", "Expected warn-limit to be a number")
    assert(cb, "A callback function must be provided when adding a warnlimit")
  end
  if tkilllimit ~= nil then
    assert(type(tkilllimit) == "number", "Expected kill-limit to be a number")
    if twarnlimit then
      assert(tkilllimit>twarnlimit, "The warnlimit must be smaller than the killlimit")
    end
  end
  if cb ~= nil then
    assert(type(cb) == "function", "Expected callback to be a function")
  end
  if hookcount ~= nil then
    assert(type(hookcount) == "number", "Expected hookcount to be a number")
    assert(hookcount >= 1, "The hookcount cannot be less than 1")
  end
  createwatch(coro, tkilllimit, twarnlimit, cb, hookcount)
  return coro
end
watch = M.watch

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.create`, except that when the running
-- coroutine is watched, then children spawned will also be watched with the same
-- settings.
-- @param f see `coroutine.create`
M.create = function(f)
  local s = getwatch(cororunning())
  if not s then return corocreate(f) end  -- I'm not being watched
  -- create and add watch
  return watch(corocreate(f), s.tkilllimit, s.twarnlimit, s.cb, s.hookcount)
end
create = M.create

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.wrap`, except that when the running
-- coroutine is watched, then children spawned will also be watched with the same
-- settings. To set sepecific settings for watching use `coroutine.wrapf`.
-- @param f see `coroutine.wrap`
-- @see wrapf
M.wrap = function(f)
  if not getwatch(cororunning()) then return corowrap(f) end  -- not watched
  local coro = create(f)
  return function(...) return resume(coro, ...) end
end

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.wrap`, except that the coroutine created
-- is watched according to the parameters provided, and not according to the watch
-- parameters of the currently running coroutine.
-- @tparam function f function to wrap
-- @tparam number|nil tkilllimit see `watch`
-- @tparam number|nil twarnlimit see `watch`
-- @tparam function|nil cb see `watch`
-- @tparam[opt=10000] number hookcount see `watch`
-- @see create
-- @see wrap
M.wrapf = function(f, tkilllimit, twarnlimit, cb, hookcount)
  local coro = watch(corocreate(f), tkilllimit, twarnlimit, cb, hookcount)
  return function(...) return resume(coro, ...) end
end

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.resume`.
-- @param coro see `coroutine.resume`
-- @param ... see `coroutine.resume`
M.resume = function(coro, ...)
  assert(type(coro) == "thread", "Expected thread, got "..type(coro))
  local e = getwatch(coro)
  if e then
    if e.errmsg then
      -- the coro being resumed is tagged with an error. This means that somewhere the error
      -- was thrown, but the coro didn't die (probably a pcall in between). The coro should have died
      -- but didn't so it is in an 'undetermined' state. Return error, don't resume
      return false, e.errmsg
    end
    local t = M.gettime()
    if e.tkilllimit then e.killtime = t + e.tkilllimit end
    if e.twarnlimit then e.warntime = t + e.twarnlimit end
    e.warned = nil
  end
  local r = pack(cororesume(coro, ...))
  if e and e.errmsg then
    return false, e.errmsg
  else
    return unpack(r)
  end
end
resume = M.resume

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.yield`.
-- @param ... see `coroutine.yield`
M.yield = function(...)
  local e = getwatch()
  if e then
    if e.errmsg then
      -- the coro is yielding, while it is tagged with an error. This means that somewhere the error
      -- was thrown, but the coro didn't die (probably a pcall in between). The coro should have died
      -- but didn't so it is in an 'undetermined' state. So kill again.
      error(e.errmsg,2)
    end
    e.killtime = nil
    e.warntime = nil
    e.warned = nil
  end
  return coroyield(...)
end

---------------------------------------------------------------------------------------
-- This is the same as the regular `coroutine.status`.
-- @param coro see `coroutine.status`
M.status = function(coro)
  if (getwatch(coro) or {}).errmsg then
    return "dead"
  else
    return corostatus(coro)
  end
end

---------------------------------------------------------------------------------------
-- This is the same as the regular `debug.sethook`, except that when trying to set a
-- hook on a coroutine that is being watched, if will throw an error.
-- @param coro see `debug.sethook`
-- @param ... see `debug.sethook`
M.sethook = function(coro, ...)
  if getwatch(coro) then
    error("Cannot set a debughook because corowatch is watching this coroutine", 2)
  end
  -- not watched, so do the regular thing
  return sethook(coro, ...)
end

---------------------------------------------------------------------------------------
-- Export the corowatch functions to an external table, or the global environment.
-- The functions exported are `create`, `yield`, `resume`, `status`, `wrap`, and `wrapf`. The standard
-- `coroutine.running` will be added if there is no `running` value in the table yet. So
-- basically it exports a complete `coroutine` table + `wrapf`.
-- If the provided table contains subtables `coroutine` and/or `debug` then it is assumed to
-- be a function/global environment and `sethook` will be exported as well (exports will then
-- go into the two subtables)
-- @tparam[opt] table t table to which to export the coroutine functions.
-- @return the table provided, or a new table if non was provided, with the exported functions
-- @usage
-- -- monkey patch global environment, both coroutine and debug tables
-- require("corowatch").export(_G)
M.export = function(t)
  t = t or {}
  local c, d
  assert(type(t) == "table", "Expected table, got "..type(t))
  if (type(t.debug) == "table") or type(t.coroutine == "table") then
    -- we got a global table, so monkeypatch debug and coroutine table
    d = t.debug
    c = t.coroutine
  else
    -- we got something else, just export coroutine here
    d = nil
    c = t
  end
  if type(d) == "table" then
    d.sethook = M.sethook
  end
  if type(c) == "table" then
    c.yield = M.yield
    c.create = M.create
    c.resume = M.resume
    c.running = c.running or cororunning
    c.status = M.status
    c.wrap = M.wrap
    c.wrapf = M.wrapf
  end
  return t
end

-- export some internals for testing if requested
if _TEST then  -- luacheck: ignore
  M._register = register
  M._getwatch = getwatch
end

return M
